Raven.Client.Lightweight/Listeners/IDocumentSaveChangesListener.cs

```diff
@@ -0,0 +1,20 @@
+//-----------------------------------------------------------------------
+// <copyright file="IDocumentStoreListener.cs" company="Hibernating Rhinos LTD">
+//     Copyright (c) Hibernating Rhinos LTD. All rights reserved.
+// </copyright>
+//-----------------------------------------------------------------------
+using Raven.Json.Linq;
+
+namespace Raven.Client.Listeners
+{
+    /// <summary>
+    /// Hook for users to provide additional logic on store operations
+    /// </summary>
+    /// 
+    public interface IDocumentSaveChangesListener
+    {
+        void BeforeSaveChanges(IDocumentSession session);
+
+        void AfterSaveChanges(IDocumentSession session);
+    }
```

Comment:
The context of the document but not the context of the unit of work IDocumentSession, for example, i'm extending it to be used with an extension method IDocumentSession.Publish that holds messages until commit : https://gist.github.com/valeriob/2a87c552be80cb3e521b545a5d288c23 so i have to hack around it, create a document Commit, and put on it messages and the di scope so that the store listener can reach both of them.

I'd rather store messages and DI scope on DocumentSession.Advanced.ExternalState and use it on the listener. https://gist.github.com/valeriob/3ae27778986da42c7a02c144d88ae52e, it allows to customize and extend unit of work (IDocumentSession) with few lines of code.

Another solution would be to wrap DocumentSession like this
https://gist.github.com/valeriob/5a200a6ed50d974e196a5a4e540316ec
but i really think the best one would be to have an extension point right after the commit has happened, with all the context around.

Thanks for the feedback Oren !